File: python/problems/leetcode/0287_find_the_duplicate_number/solution.py
"""LeetCode 287: Find the Duplicate Number."""

from __future__ import annotations


class Solution:
    def findDuplicate(self, nums: list[int]) -> int:
        slow = nums[0]
        fast = nums[0]
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
        slow = nums[0]
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]
        return slow


if __name__ == "__main__":
    assert Solution().findDuplicate([1, 3, 4, 2, 2]) == 2
    assert Solution().findDuplicate([3, 1, 3, 4, 2]) == 3
    assert Solution().findDuplicate([1, 1]) == 1
    print("0287 OK")

File: cpp/problems/leetcode/0287_find_the_duplicate_number/solution.cpp
// LeetCode 287
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    int findDuplicate(vector<int>& nums) {
        int slow = nums[0], fast = nums[0];
        do {
            slow = nums[slow];
            fast = nums[nums[fast]];
        } while (slow != fast);
        slow = nums[0];
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }
        return slow;
    }
};

int main() {
    vector<int> a{1, 3, 4, 2, 2};
    assert(Solution().findDuplicate(a) == 2);
    vector<int> b{3, 1, 3, 4, 2};
    assert(Solution().findDuplicate(b) == 3);
    vector<int> c{1, 1};
    assert(Solution().findDuplicate(c) == 1);
    cout << "0287 OK" << endl;
    return 0;
}
